language/tools/vm-genesis/src/lib.rs

```diff
@@ -351,18 +351,32 @@ fn create_and_initialize_owners_operators(
 /// Publish the standard library.
 fn publish_stdlib(session: &mut Session<StateViewCache>, stdlib: Modules) {
     let dep_graph = stdlib.compute_dependency_graph();
-    for module in dep_graph.compute_topological_order().unwrap() {
-        let module_id = module.self_id();
-        if module_id.name().as_str() == GENESIS_MODULE_NAME {
-            // Do not publish the Genesis module
-            continue;
-        }
-        let mut bytes = vec![];
-        module.serialize(&mut bytes).unwrap();
-        session
-            .publish_module(bytes, *module_id.address(), &mut GasStatus::new_unmetered())
-            .unwrap_or_else(|e| panic!("Failure publishing module {:?}, {:?}", module_id, e));
-    }
+    let mut addr_opt: Option<AccountAddress> = None;
+    let modules = dep_graph
+        .compute_topological_order()
+        .unwrap()
+        .map(|m| {
+            let addr = *m.self_id().address();
+            if let Some(a) = addr_opt {
+              assert!(
+                  a == addr,
+                  "All genesis modules must be published under the same address, but found modules under both {} and {}",
+                  a.short_str_lossless(),
+                  addr.short_str_lossless()
+              );
+            } else {
+                addr_opt = Some(addr)
+            }
+            let mut bytes = vec![];
+            m.serialize(&mut bytes).unwrap();
+            bytes
+        })
+        .collect::<Vec<Vec<u8>>>();
+    // TODO: allow genesis modules published under different addresses. supporting this while
+    // maintaining the topological order is challenging.
```

Comment:
```suggestion
    // TODO: allow genesis modules published under different addresses. supporting this while
    // maintaining the topological order is challenging. This requires the Move VM to support
    // multi-module publishing with multiple signers as well, whereas the current Move VM
    // publishing API allows only a single singer only.
```

Reply:
Interesting...

Given a linking-safe set of modules `S` that have some intra-address `friend` relationships (I think `friend` already has that restriction) , but with inter-address dependencies, do you think it is always possible to publish `S` via a sequence of calls to the current VM publishing API's?

My guess is yes, but I haven't tried too hard to find a counterexample + haven't attempted a proof either.

Reply:
Ahh, I think you are right. We should be able to publish S via a sequence of calls to the single-signer publishing API.

An informal argument is this:

- Consider all modules in S as nodes. There is a directional edge between two modules, A and B, if
  - A uses B in any way, either 1) calling a function in B or 2) using a type defined in B, or
  - A marks B as a friend.
  By definition, if A uses B's function (A->B) and B marks A (B->A) as a friend, this creates a loop.

- We can then partition S into strongly-connected components and topologically sort all the SCCs into [C1, C2, ..., Cn], where each Ci could be either a single module or a loop.
  - If an SCC contains a single module, we could just publish that single module
  - If an SCC contains a loop, all modules in that loop must be published together as a bundle.
    And the single-address friendship requires that all modules in the loop must share the same address.

And hence, we are fine with just a single-signer publishing, as long as the single-address friendship requirement is in place.

Reply:
Nice, thanks for the detailed analysis!!

I think we *might* still want a multi-signer publishing API for a different reason--you could imagine wanting to atomically update modules published under different addresses. But I don't think we have a strong motivation for that right now.